### PR TITLE
Setup initial release of Python package

### PR DESCRIPTION
**Pull request by nicoleepp**
This configures this repo as a python package so that we can use its `Changelog` code in other python projects.

### CODE REVIEW

File: docs/general/gen_template.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+## Template
```
Comment (fchorney) [inline docs/general/gen_template.md:1]: don't need this file

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+[tool.black]
+target-version = ["py38"]
```
Comment (fchorney) [inline pyproject.toml:2]:
```suggestion
target-version = ["py37", "py38"]
```

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import codecs
+from os import path
+
+from setuptools import find_packages, setup
+
+
+TEST_DEPS = ["coverage", "pytest", "pytest-cov", "regex"]
```
Comment (fchorney) [inline setup.py:7]: What is `regex`?

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    url="https://github.com/invenia/tagbotgitlab",
+    packages=find_packages(exclude=["tests"]),
+    install_requires=REQUIREMENTS,
+    classifiers=[
+        "Environment :: Console",
+        "Intended Audience :: Developers",
+        "License :: OSI Approved :: MIT License",
+        "Operating System :: OS Independent",
+        "Programming Language :: Python",
+        "Programming Language :: Python :: 3.8",
```
Comment (fchorney) [inline setup.py:40]:
```suggestion
        "Programming Language :: Python :: 3.7",
        "Programming Language :: Python :: 3.8",
```

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    py38,
+    coverage
+    docs
+
+[testenv]
+usedevelop = True
+extras = test
+setenv =
+    PYTHONPATH = {toxinidir}
+    py{37,38}: COVERAGE_FILE=.coverage.{envname}
```
Comment (fchorney) [inline tox.ini:14]:
```suggestion
    COVERAGE_FILE=.coverage.{envname}
```
Turns out we don't need that here

File: tox.ini [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+commands =
+    py{37,38}: py.test --cov=tagbotgitlab --verbose --tb=long
```
Comment (fchorney) [inline tox.ini:16]:
```suggestion
    py.test --cov=tagbotgitlab --verbose --tb=long
```
also here

**Comment by fchorney** [comment]
If you want, you can try using https://packaging.python.org/guides/using-testpypi/ to do a test upload to the test pypi server, to make sure everything is as you expect, etc

File: tests/test_version.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+def test_version_string():
+    # Regex found here: https://github.com/k-bx/python-semver/blob/master/semver.py
+    regex = (
+        r"^(?P<major>(?:0|[1-9][0-9]*))\."
+        r"(?P<minor>(?:0|[1-9][0-9]*))\."
+        r"(?P<patch>(?:0|[1-9][0-9]*))"
+        r"(\-(?P<prerelease>(?:0|[1-9A-Za-z-][0-9A-Za-z-]*)"
+        r"(\.(?:0|[1-9A-Za-z-][0-9A-Za-z-]*))*))?"
+        r"(\+(?P<build>[0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*))?$"
+    )
```
Comment (omus) [inline tests/test_version.py:16]: Consider just using `VERBOSE` here. I also pulled the official regex as provided by: https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
```python
regex = re.compile(r"""
    ^(?P<major>0|[1-9]\d*)\.
    (?P<minor>0|[1-9]\d*)\.
    (?P<patch>0|[1-9]\d*)
    (?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)
    (?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?
    (?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$
""", re.VERBOSE)
```

Comment (fchorney) [reply tests/test_version.py:16]: Looks like the same regex to me? Except `\d` is being replaced by `[0-9]`
the VERBOSe is a good idea. I should add that to the cookiecutter

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import codecs
+from os import path
+
+from setuptools import find_packages, setup
+
+
+TEST_DEPS = ["coverage", "pytest", "pytest-cov"]
+DOCS_DEPS = ["sphinx", "sphinx-rtd-theme", "sphinx-autoapi", "recommonmark"]
+CHECK_DEPS = ["isort", "flake8", "flake8-quotes", "pep8-naming", "mypy", "black"]
+REQUIREMENTS = ["Jinja2", "python-dateutil", "python-gitlab", "semver"]
```
[the remark this answers is not in the document]
Comment (fchorney) [reply setup.py:10]: just fyi for how to specify lower/upper bounds https://packaging.python.org/discussions/install-requires-vs-requirements/

**Comment by nicoleepp** [comment]
I have tested uploading this to testpypi

File: tests/test_version.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+        r"""
+    ^(?P<major>0|[1-9]\d*)\.
+    (?P<minor>0|[1-9]\d*)\.
+    (?P<patch>0|[1-9]\d*)
+    (?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)
+    (?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?
+    (?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$
+""",
+        re.VERBOSE,
+    )
```
Comment (fchorney) [inline tests/test_version.py:19]:
```suggestion
def test_version_string():
    # Regex found here: https://github.com/k-bx/python-semver/blob/master/semver.py
    regex = re.compile(
        r"""
            ^(?P<major>0|[1-9]\d*)\.
            (?P<minor>0|[1-9]\d*)\.
            (?P<patch>0|[1-9]\d*)
            (?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)
            (?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?
            (?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$
        """,
        re.VERBOSE,
    )
```
Looks like you can format this a little nicer and black is still happy

Comment (nicoleepp) [reply tests/test_version.py]: You are right!

**Pull request merged**